exercises/circular-buffer/src/test/java/CircularBufferTest.java

```diff
@@ -0,0 +1,162 @@
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class CircularBufferTest {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @Test
+    public void readingFromEmptyBufferShouldThrowException() throws BufferIOException {
+        CircularBuffer<Integer> buffer = new CircularBuffer<>(1);
+
+        expectedException.expect(BufferIOException.class);
+        expectedException.expectMessage("Tried to read from empty buffer");
+        buffer.read();
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void canReadItemJustWritten() throws BufferIOException {
+        CircularBuffer<Integer> buffer = new CircularBuffer<>(1);
+
+        buffer.write(1);
+        assertEquals((int) buffer.read(), 1);
```

Comment:
@FridaTveit, the int cast was indeed due to the error that @Smarticles101 posted. I wrestled with that int cast for a couple hours & just now figured out a) why it does that and b) how to avoid it! It seems to be due to some wackiness with how Java autoboxing & unboxing interacts with target type inference.

The problem goes like this: Since Java `Collections` can't store raw types, any method we use to pull values out of our data structure will return `Integer` rather than `int`. So, my line `return buffer.remove()` returns an `Integer`. This then becomes an argument to `assertEquals()`, but the other argument is an `int`. `assertEquals()` is overloaded, but doesn't have any implementation with the type signature `Integer, int`, so it's stuck deciding between its two best guesses for the type signature: `long, long` and `Object, Object`.

So if we stick with `assertEquals()`, we can either: 1) remove the cast, in which case any users trying to implement the exercise with a `Collection` may run into this confusing bug, or 2) kludge both the arguments to the same type with a cast so they're both unboxed or both boxed.

But, `assertThat()` to the rescue! Turns out that `asssertThat(Integer i, is(int j))` is unambiguously typed & neatly sidesteps the problem. So I redid the tests with that & we should be golden, I think.